### ISSUE TITLE
How to dispatch a new action after a successful update

### ISSUE THREAD

**Issue by imbradyboy** [ngxs-labs/firestore-plugin#37]
I've gone through the resources and documentation and can't seem to find exactly what I'm looking for.

I have an array (of objects) in a Firestore document, and after an item has been added to this array (via arrayUnion) I want to dispatch a routing event to take you to the name of this new array element. I need to wait for the update operation to complete before I do this, so how do I monitor the completion (success or error) of operations using this plugin?

**Comment by skotturi** [comment]
Using the `StreamEmitted` method this plugin provides. If you post some of your code, it might help

**Comment by imbradyboy** [comment]
> Using the `StreamEmitted` method this plugin provides. If you post some of your code, it might help

`      const state = ctx.getState();
      
      // // this is a create so add a new project to array
      // TODO need to catch errors
      const update$ = await this.update$(this.auth.getUserID(), {
        projects: firebase.default.firestore.FieldValue.arrayUnion(newProject)
      }, {merge: true});
      
      // TODO need to notify front end when this is completed
      await update$.pipe(take(1)).subscribe(projectId => {
        ctx.dispatch(new SetSelectedProject(projectId));
      });`

Basically, I need a way to notify the front end component when everything has completed successfully/errors

**Comment by joaqcid** [comment]
@imbradyboy you have different ways to do this.

if you are wrapping your update within an action you can then subscribe to it, to execute after it was completed.
in order to get the value of the last inserted value, you can `selectSnapshot` and then use it to navigate to it

you would have in your state something like
```
@Action(Update)
async update(ctx: StateContext<>, action: Update){
 await this.firestore.update$(this.auth.getUserID(), {
    projects: firebase.default.firestore.FieldValue.arrayUnion(newProject)
  }, {merge: true});
  // save newProject in state after was saved in firestore
  ctx.patchState()
}
```

and in your component

```
this.store.dispatch(Update).subscribe(()=>{
  const newProject = this.store.selectSnapshot(...new project selector)
  this.router.navigateToUrl(...)
})
```

hope this helps

**Comment by imbradyboy** [comment]
> @imbradyboy you have different ways to do this.
> 
> if you are wrapping your update within an action you can then subscribe to it, to execute after it was completed.
> in order to get the value of the last inserted value, you can `selectSnapshot` and then use it to navigate to it
> 
> you would have in your state something like
> 
> ```
> @Action(Update)
> async update(ctx: StateContext<>, action: Update){
>  await this.firestore.update$(this.auth.getUserID(), {
>     projects: firebase.default.firestore.FieldValue.arrayUnion(newProject)
>   }, {merge: true});
>   // save newProject in state after was saved in firestore
>   ctx.patchState()
> }
> ```
> 
> and in your component
> 
> ```
> this.store.dispatch(Update).subscribe(()=>{
>   const newProject = this.store.selectSnapshot(...new project selector)
>   this.router.navigateToUrl(...)
> })
> ```
> 
> hope this helps

Thank you! I was able to figure it out. I do have a question though (new to ngxs so forgive me if it's a dumb question). Why do the Firestore update functions in your service return observables instead of Promises? I thought they would only emit once (resolve or reject)

**Comment by joaqcid** [comment]
great! 

we decided to use observables for write operations  (create, update, delete) so all methods are observables. even though `update$` returns an observables it will only emit once after it has completed, so in practice it will behave same as a Promise
